ADBDEV-6460 Fix -Wvla-cxx-extension warning

dnskvlnk opened the pull request:
Fix -Wvla-cxx-extension warning produced by Clang

dnskvlnk committed:
-Wvla-cxx-extension

dnskvlnk committed:
Fix orca formatting

andr-sokolov commented on src/backend/gpopt/utils/COptTasks.cpp:
```diff
@@ -987,13 +987,13 @@ COptTasks::PrintMissingStatsWarning(CMemoryPool *mp, CMDAccessor *md_accessor,
 	{
 		int length = NAMEDATALEN * rel_stats->Size() + 200;
-		char msgbuf[length];
+		std::vector<char> msgbuf(length);
```
What about using some string class instead of vector? For example std::string

dnskvlnk replied:
I'm not sure that it's safe to pass std::string::data() like I did here

andr-sokolov replied:
What about something like
```c
if (0 == ul)
{
    oss << "One or more columns in the following table(s) do not have statistics: ";
}
else
{
    oss << ", ";
}
```
?

andr-sokolov commented on src/backend/gpopt/translate/CTranslatorUtils.cpp:
```diff
@@ -387,8 +387,8 @@ CTranslatorUtils::ResolvePolymorphicTypes(CMemoryPool *mp,
-	OID arg_types[total_args];
-	char arg_modes[total_args];
+	std::vector<OID> arg_types(total_args);
+	std::vector<char> arg_modes(total_args);
```
It will be better to use some string here too

dnskvlnk replied:
I think that vector is more suitable here because usage of string implies that you actually imply a string. This is an array of chars not related to each other

andr-sokolov commented:
The description contains two identical sentences. They can be combined in one "Fix -Wvla-cxx-extension warning produced by Clang". The first letter in the Clang word should be in upper case

bandetto commented on src/backend/gpopt/utils/COptTasks.cpp:
```diff
@@ -987,13 +987,13 @@ COptTasks::PrintMissingStatsWarning(CMemoryPool *mp, CMDAccessor *md_accessor,
+		std::vector<char> msgbuf(length);
 		snprintf(
-			msgbuf, sizeof(msgbuf),
+			msgbuf.data(), msgbuf.size() * sizeof(char),
```
Vector already has an element size in template arguments.
```suggestion
			msgbuf.data(), msgbuf.size(),
```

dnskvlnk replied:
No, size return the number of elements currently held by a vector

bandetto replied:
Sorry, forgot that sprintf() is from C.

andr-sokolov commented:
Orca almost does not use vector, string and other similar classes from standard library, because Orca allocates memory using memory pools. What about using memory pool directly or using classes like CWStringDynamic, which allocate memory in pool?

dnskvlnk committed:
Use ORCA's CWStringDynamic instead of std::string

This commit substitutes std::string with ORCA's CWStringDynamic because ORCA
extensively uses wchar_t instead of the regular char, which forces to use
std::wstring. However, std::wstring doesn't have a constructor which takes
an array of chars/pointer to chars. Therefore, alternate use of std::string and
std::wstring is complicated and takes more lines than CWStringDynamic.

Commit:
Fix orca formatting

andr-sokolov commented on src/backend/gpopt/utils/COptTasks.cpp:
```diff
@@ -986,17 +986,20 @@ COptTasks::PrintMissingStatsWarning(CMemoryPool *mp, CMDAccessor *md_accessor,
-		snprintf(
-			msgbuf, sizeof(msgbuf),
-			"One or more columns in the following table(s) do not have statistics: %s",
+		CWStringDynamic *msgbuf = GPOS_NEW(mp) CWStringDynamic(mp);
```
It is redundant to create msgbuf, because we need one string only. The wcstr string is not used without "One or more columns in the following table(s) do not have statistics: ". So it will be more optimal to begin wcstr with "One or more columns in the following table(s) do not have statistics: ". In this case the table list will not be copied from one string to another.

dnskvlnk replied:
I think it's a bad practice to append diagnostic messages into proper names and I would leave it as is

andr-sokolov replied:
wcstr is used in diagnostic message only. Bad practice is writing complex code, allocating more memory and consuming more CPU time than possible. Why is my suggestion bad?

dnskvlnk replied:
How much faster will the program work if I make this insignificant change? I don't have a problem changing it if it increases the speed dramatically. As for the complexity, complexity is filling a string with the content not expected in it. There's a reason the original code had a dedicated buffer for it

andr-sokolov replied:
@bandetto, Do you agree that msgbuf is redundant?

dnskvlnk replied:
You pose the question in a way that makes it easier for the reviewer to agree with you. The current solution is implemented in accordance with the style and technical approach of the code base. The solution you propose is based on your personal preferences, not on objective arguments. You left the question about increasing the execution time unanswered. All proposed changes should be supported by technical arguments, not personal preferences.

bandetto replied:
> Do you agree that msgbuf is redundant?

It's redundant, but I'd choose readability here anyway, since this likely isn't a hot path that really matters for performance.

andr-sokolov replied:
Ok. Let's create msgbuf on stack like wcstr

Commit:
Merge branch 'adb-7.2.0' into ADBDEV-6460

Commit:
Merge branch 'adb-7.2.0' into ADBDEV-6460

Commit:
Allocate msgbuf on stack

Commit:
Merge branch 'adb-7.2.0' into ADBDEV-6460

andr-sokolov commented on src/backend/gpopt/utils/COptTasks.cpp:
```diff
@@ -986,14 +986,15 @@ COptTasks::PrintMissingStatsWarning(CMemoryPool *mp, CMDAccessor *md_accessor,
+		CWStringDynamic msgbuf(mp);
+		msgbuf.AppendFormat(
+			GPOS_WSZ_LIT(
+				"One or more columns in the following table(s) do not have statistics: %s"),
```
msgbuf is string of WCHAR and wcstr is string of WCHAR too. So we can replace "%s" with "%ls" and don't call CreateMultiByteCharStringFromWCString

dnskvlnk replied:
Done

Commit:
Simplify msgbuf concatenation

Commit:
Merge branch 'adb-7.2.0' into ADBDEV-6460
